File: src/main/scala/com/projextxy/core/client/gui/GuiRGB.java
package com.projextxy.core.client.gui;

import codechicken.lib.colour.ColourRGBA;
import com.projextxy.core.ProjectXYCore$;
import com.projextxy.core.inventory.ContainerRGB;
import com.projextxy.core.reference.MCColors;
import com.projextxy.core.tile.TileColorizer;
import com.projextxy.lib.cofh.gui.GuiColor;
import com.projextxy.lib.cofh.gui.element.listbox.SliderHorizontal;
import net.minecraft.entity.player.InventoryPlayer;
import net.minecraft.util.ResourceLocation;


public class GuiRGB extends GuiCommon {
    private final TileColorizer tileCustomColor;

    public GuiRGB(InventoryPlayer inventory, TileColorizer tileColorizer) {
        super(new ContainerRGB(inventory, tileColorizer), new ResourceLocation(ProjectXYCore$.MODULE$.MOD_ID().toLowerCase(), "/textures/gui/guiColorizer.png"));
        this.tileCustomColor = tileColorizer;
        drawInventory = false;
    }

    @Override
    public void initGui() {
        super.initGui();
        SliderHorizontal sliderRed = new SliderHorizontal(this, 10, 20, xSize - 20, 15, 255) {
            @Override
            public void onValueChanged(int value) {
                tileCustomColor.setR(value);
                tileCustomColor.buildPacket().sendToServer();
            }
        };
        SliderHorizontal sliderGreen = new SliderHorizontal(this, 10, 45, xSize - 20, 15, 255) {
            @Override
            public void onValueChanged(int value) {
                tileCustomColor.setG(value);
                tileCustomColor.buildPacket().sendToServer();
            }
        };

        SliderHorizontal sliderBlue = new SliderHorizontal(this, 10, 70, xSize - 20, 15, 255) {
            @Override
            public void onValueChanged(int value) {
                tileCustomColor.setB(value);
                tileCustomColor.buildPacket().sendToServer();
            }
        };

        sliderRed.setValue(tileCustomColor.getR());
        sliderGreen.setValue(tileCustomColor.getG());
        sliderBlue.setValue(tileCustomColor.getB());
        sliderRed.setColor(new GuiColor(120, 120, 120, 255).getColor(), new GuiColor(60, 60, 60, 255).getColor());
        sliderGreen.setColor(new GuiColor(120, 120, 120, 255).getColor(), new GuiColor(60, 60, 60, 255).getColor());
        sliderBlue.setColor(new GuiColor(120, 120, 120, 255).getColor(), new GuiColor(60, 60, 60, 255).getColor());
        elements.add(sliderRed);
        elements.add(sliderGreen);
        elements.add(sliderBlue);
    }

    @Override
    protected void drawGuiContainerForegroundLayer(int x, int y) {
        super.drawGuiContainerForegroundLayer(x, y);
        fontRendererObj.drawStringWithShadow("Red: " + String.valueOf(tileCustomColor.getR()), 10, 10, MCColors.RED.rgb);
        fontRendererObj.drawStringWithShadow("Green: " + String.valueOf(tileCustomColor.getG()), 10, 36, MCColors.GREEN.rgb);
        fontRendererObj.drawStringWithShadow("Blue: " + String.valueOf(tileCustomColor.getB()), 10, 61, MCColors.BLUE.rgb);
        drawProgressBar(xSize / 2 - 14, 93, 1D, new ColourRGBA(tileCustomColor.getR(), tileCustomColor.getG(), tileCustomColor.getB(), 255));
    }

    @Override
    public void drawBackground(int p_146278_1_) {
        super.drawBackground(p_146278_1_);
    }

    @Override
    public void onGuiClosed() {
        tileCustomColor.buildPacket().sendToServer();
        super.onGuiClosed();
    }
}

File: src/main/scala/com/projextxy/core/client/gui/GuiRGB.scala
package com.projextxy.core.client.gui

import codechicken.lib.colour.ColourRGBA
import com.projextxy.core.inventory.ContainerRGB
import com.projextxy.core.reference.MCColors
import com.projextxy.core.tile.TileColorizer
import net.minecraft.client.gui.GuiTextField
import net.minecraft.entity.player.InventoryPlayer
import org.lwjgl.opengl.GL11

class GuiRGB(inventory: InventoryPlayer, tile: TileColorizer) extends GuiCommon(new ContainerRGB(inventory, tile)) {

  var rText: GuiTextField = null
  var gText: GuiTextField = null
  var bText: GuiTextField = null

  override def initGui(): Unit = {
    super.initGui()
    rText = new GuiTextField(fontRendererObj, guiLeft + 8, guiTop + 45, 30, 10)
    gText = new GuiTextField(fontRendererObj, guiLeft + xSize / 2 - 15, guiTop + 45, 30, 10)
    bText = new GuiTextField(fontRendererObj, guiLeft + xSize - 38, guiTop + 45, 30, 10)
    rText.setText(tile.getR.toString)
    gText.setText(tile.getG.toString)
    bText.setText(tile.getB.toString)
  }


  override def mouseClicked(mouseX: Int, mouseY: Int, button: Int): Unit = {
    super.mouseClicked(mouseX, mouseY, button)
    rText.mouseClicked(mouseX, mouseY, button)
    gText.mouseClicked(mouseX, mouseY, button)
    bText.mouseClicked(mouseX, mouseY, button)
  }


  override def keyTyped(key: Char, i: Int): Unit = {
    super.keyTyped(key, i)

    def doTextChange(guiTextField: GuiTextField): Boolean = {
      if (guiTextField.isFocused) if (guiTextField.getText.length < 3)
        if ((guiTextField.getText + key).toInt <= 255) {
          guiTextField.textboxKeyTyped(key, i)
          return true
        }
      false
    }

    if (key.isDigit) {
      if (doTextChange(rText)) {
        tile.setR(getValueFromText(rText))
        tile.buildPacket().sendToServer()
      }
      if (doTextChange(gText)) {
        tile.setG(getValueFromText(gText))
        tile.buildPacket().sendToServer()
      }
      if (doTextChange(bText)) {
        tile.setB(getValueFromText(bText))
        tile.buildPacket().sendToServer()
      }
    } else if (key.isControl) {
      rText.textboxKeyTyped(key, i)
      gText.textboxKeyTyped(key, i)
      bText.textboxKeyTyped(key, i)
    }
  }

  def getValueFromText(textField: GuiTextField) = {
    try {
      textField.getText.toInt
    } catch {
      case numberFormat: NumberFormatException => 0
    }
  }

  override def drawScreen(mouseX: Int, mouseY: Int, par3: Float): Unit = {
    super.drawScreen(mouseX, mouseY, par3)
    GL11.glDisable(GL11.GL_LIGHTING)
    rText.drawTextBox()
    gText.drawTextBox()
    bText.drawTextBox()
    GL11.glEnable(GL11.GL_LIGHTING)
  }

  override def updateScreen(): Unit = {
    super.updateScreen()
  }

  override def onGuiClosed(): Unit = {
    tile.buildPacket().sendToServer()
    tile.getWorldObj.markBlockForUpdate(tile.xCoord, tile.yCoord, tile.zCoord)
    super.onGuiClosed()
  }

  override def guiName: String = "Color Injector"

  override def drawBackground(): Unit = {
    drawProgressBar(xSize / 2 - 15, 70, 1.0D, new ColourRGBA(getValueFromText(rText), getValueFromText(gText), getValueFromText(bText), 255))
  }

  override def drawForeground(): Unit = {
    fontRendererObj.drawStringWithShadow("Red", 8, 35, MCColors.RED.rgb)
    fontRendererObj.drawStringWithShadow("Green", xSize / 2 - 15, 35, MCColors.GREEN.rgb)
    fontRendererObj.drawStringWithShadow("Blue", xSize - 38, 35, MCColors.BLUE.rgb)

    tile.setR(getValueFromText(rText))
    tile.setG(getValueFromText(gText))
    tile.setB(getValueFromText(bText))
  }
}
